### PR TITLE
Python formatting tools

### PR DESCRIPTION
**Pull request by Xierumeng**
Added [Black formatter](https://black.readthedocs.io/en/stable/index.html) and addressed all [Pylint linter](https://pylint.pycqa.org/en/stable/index.html) issues.

### CODE REVIEW

**Review by mgupta27 — APPROVED** [review]
LGTM

File: documentation/tests/unit/test_add_or_multiply.py [hunk 1]
```diff
@@ -5,115 +5,123 @@
...
 # Makes it easier when setup is complicated
 @pytest.fixture()
-def adder():
+def adder() -> add_or_multiply.AddOrMultiply:  # type: ignore
     """
     Creates AddOrMultiply in addition state.
     """
     add = add_or_multiply.AddOrMultiply(add_or_multiply.MathOperation.ADD)
-    yield add
+    yield add  # type: ignore
```
Comment (Xierumeng) [inline documentation/tests/unit/test_add_or_multiply.py:29]: Is ignore necessary?

File: documentation/tests/unit/test_pytest_example.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Pytest example. To run:
+```
+cd documentation/tests/
+pytest
+```
+"""
+
+import pytest
+
+
```
Comment (Xierumeng) [inline documentation/tests/unit/test_pytest_example.py:11]: Add pylint disables.

File: modules/cluster_estimation/cluster_estimation.py [hunk 3]
```diff
@@ -303,21 +311,23 @@ def __filter_by_points_ownership(self,
         filtered_output: list[tuple[np.ndarray, float, float]]
             List containing predicted cluster centres after filtering.
         """
-
-        results = self.__vgmm.predict(self.__all_points)
+        results = self.__vgmm.predict(self.__all_points)  # type: ignore
         filtered_output = []
```
Comment (Xierumeng) [inline modules/cluster_estimation/cluster_estimation.py:315]: Move and annotate.

File: modules/cluster_estimation/cluster_estimation.py [hunk 4]
```diff
@@ -303,21 +311,23 @@ def __filter_by_points_ownership(self,
...
 
         # Filtering by each cluster's point ownership
-        unique_clusters, num_points_per_cluster = np.unique(results, return_counts=True)
+        unique_clusters, _ = np.unique(results, return_counts=True)
         # Remove empty clusters
+        # Integer index required
+        # pylint: disable-next=consider-using-enumerate
         for i in range(len(model_output)):
             if i in unique_clusters:
                 filtered_output.append(model_output[i])
```
Comment (Xierumeng) [inline modules/cluster_estimation/cluster_estimation.py:324]: Change to list comprehension.

File: modules/detections_and_time.py [hunk 5]
```diff
@@ -37,61 +35,67 @@ def create(cls,
...
+            + str(self.confidence)
+            + ", bounds: "
+            + str(self.x_1)
+            + " "
+            + str(self.y_1)
+            + " "
+            + str(self.x_2)
+            + " "
+            + str(self.y_2)
+        )
```
Comment (TongguangZhang) [inline modules/detections_and_time.py:71]: Use f-strings

File: modules/detections_and_time.py [hunk 6]
```diff
@@ -115,10 +119,16 @@ def __init__(self, class_private_create_key, timestamp: float):
...
+        """
+        String representation.
+        """
+        representation = (
+            str(self.__class__)
+            + ", time: "
+            + str(int(self.timestamp))
+            + ", size: "
+            + str(len(self))
+        )
```
Comment (TongguangZhang) [inline modules/detections_and_time.py:131]: Use f-strings

File: modules/drone_odometry_local.py [hunk 7]
```diff
@@ -37,24 +33,33 @@ def __init__(self,
...
+        String representation.
+        """
+        return (
+            "DronePositionLocal (NED): "
+            + str(self.north)
+            + ", "
+            + str(self.east)
+            + ", "
+            + str(self.down)
+        )
```
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:47]: Use f-strings

File: modules/drone_odometry_local.py [hunk 8]
```diff
@@ -66,36 +71,50 @@ def create_new(cls,
...
+        """
+        # TODO: Update common
+        return (
+            "DroneOrientationLocal (YPR): "
+            + str(self.orientation.yaw)
+            + ", "
+            + str(self.orientation.pitch)
+            + ", "
+            + str(self.orientation.roll)
+        )
```
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:104]: Use f-strings

File: modules/geolocation/camera_properties.py [hunk 9]
```diff
@@ -41,26 +41,26 @@ def create_rotation_matrix_from_orientation(yaw: float,
     yaw_matrix = np.array(
         [
             [np.cos(yaw), -np.sin(yaw), 0.0],
-            [np.sin(yaw),  np.cos(yaw), 0.0],
-            [        0.0,          0.0, 1.0],
+            [np.sin(yaw), np.cos(yaw), 0.0],
+            [0.0, 0.0, 1.0],
```
Comment (Xierumeng) [inline modules/geolocation/camera_properties.py:45]: UNDO.

Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:45]: https://stackoverflow.com/questions/58584413/black-formatter-ignore-specific-multi-line-code

Comment (TongguangZhang) [inline modules/geolocation/camera_properties.py:45]: Add formatting ignore

File: modules/geolocation/camera_properties.py [hunk 10]
```diff
@@ -41,26 +41,26 @@ def create_rotation_matrix_from_orientation(yaw: float,
...
         dtype=np.float32,
     )
 
     pitch_matrix = np.array(
         [
-            [ np.cos(pitch), 0.0, np.sin(pitch)],
-            [           0.0, 1.0,           0.0],
+            [np.cos(pitch), 0.0, np.sin(pitch)],
+            [0.0, 1.0, 0.0],
             [-np.sin(pitch), 0.0, np.cos(pitch)],
```
Comment (TongguangZhang) [inline modules/geolocation/camera_properties.py:54]: Add formatting ignore

File: modules/geolocation/camera_properties.py [hunk 11]
```diff
@@ -41,26 +41,26 @@ def create_rotation_matrix_from_orientation(yaw: float,
...
         dtype=np.float32,
     )
 
     roll_matrix = np.array(
         [
-            [1.0,          0.0,           0.0],
+            [1.0, 0.0, 0.0],
             [0.0, np.cos(roll), -np.sin(roll)],
-            [0.0, np.sin(roll),  np.cos(roll)],
+            [0.0, np.sin(roll), np.cos(roll)],
```
Comment (TongguangZhang) [inline modules/geolocation/camera_properties.py:63]: Add formatting ignore

File: modules/geolocation/camera_properties.py [hunk 12]
```diff
@@ -117,28 +116,30 @@ def create(cls,
...
 
         self.resolution_x = resolution_x
         self.resolution_y = resolution_y
 
-        self.__vec_c = np.array([1.0,      0.0,      0.0], dtype=np.float32)
-        self.__vec_u = np.array([0.0, u_scalar,      0.0], dtype=np.float32)
-        self.__vec_v = np.array([0.0,      0.0, v_scalar], dtype=np.float32)
+        self.__vec_c = np.array([1.0, 0.0, 0.0], dtype=np.float32)
+        self.__vec_u = np.array([0.0, u_scalar, 0.0], dtype=np.float32)
+        self.__vec_v = np.array([0.0, 0.0, v_scalar], dtype=np.float32)
```
Comment (TongguangZhang) [inline modules/geolocation/camera_properties.py:137]: Add formatting ignore

File: modules/merged_odometry_detections.py [hunk 13]
```diff
@@ -28,17 +28,29 @@ def create(cls,
...
         self.detections = detections
 
-# pylint: enable=too-few-public-methods
+    def __repr__(self) -> str:
+        """
+        String representation.
+        """
+        representation = (
+            "Merged: " + str(self.odometry_local) + ", detections: " + str(len(self.detections))
+        )
```
Comment (TongguangZhang) [inline modules/merged_odometry_detections.py:53]: Use f-string

File: modules/odometry_and_time.py [hunk 14]
```diff
@@ -40,4 +43,11 @@ def __init__(self,
...
         self.timestamp = timestamp
 
-# pylint: enable=too-few-public-methods
+    def __repr__(self) -> str:
+        """
+        String representation.
+        """
+        representation = str(self.__class__) + ", time: " + str(int(self.timestamp))
+
+        representation += "\n" + repr(self.odometry_data)
```
Comment (TongguangZhang) [inline modules/odometry_and_time.py:52]: Use f-string

File: tests/unit/test_camera_properties.py [hunk 15]
```diff
@@ -209,11 +220,11 @@ def test_pitch_quarter(self):
 
         expected = np.array(
             [
-                [ 0.0, 0.0, 1.0],
-                [ 0.0, 1.0, 0.0],
+                [0.0, 0.0, 1.0],
+                [0.0, 1.0, 0.0],
                 [-1.0, 0.0, 0.0],
```
Comment (TongguangZhang) [inline tests/unit/test_camera_properties.py:225]: Add formatter ignore

File: tests/unit/test_camera_properties.py [hunk 16]
```diff
@@ -239,11 +250,11 @@ def test_roll_quarter(self):
 
         expected = np.array(
             [
-                [1.0, 0.0,  0.0],
+                [1.0, 0.0, 0.0],
                 [0.0, 0.0, -1.0],
-                [0.0, 1.0,  0.0],
+                [0.0, 1.0, 0.0],
```
Comment (TongguangZhang) [inline tests/unit/test_camera_properties.py:255]: Add formatter ignore

File: tests/unit/test_camera_properties.py [hunk 17]
```diff
@@ -269,11 +280,11 @@ def test_combined_rotations_positive(self):
 
         expected = np.array(
             [
-                [          1 / 2, (np.sqrt(2) - 2) / 4, (np.sqrt(2) + 2) / 4],
-                [          1 / 2, (np.sqrt(2) + 2) / 4, (np.sqrt(2) - 2) / 4],
-                [-np.sqrt(2) / 2,                1 / 2,                1 / 2],
+                [1 / 2, (np.sqrt(2) - 2) / 4, (np.sqrt(2) + 2) / 4],
+                [1 / 2, (np.sqrt(2) + 2) / 4, (np.sqrt(2) - 2) / 4],
+                [-np.sqrt(2) / 2, 1 / 2, 1 / 2],
```
Comment (TongguangZhang) [inline tests/unit/test_camera_properties.py:285]: Add formatter ignore

File: tests/unit/test_camera_properties.py [hunk 18]
```diff
@@ -299,11 +310,11 @@ def test_combined_rotations_negative(self):
 
         expected = np.array(
             [
-                [         1 / 2,  (np.sqrt(2) + 2) / 4, (-np.sqrt(2) + 2) / 4],
-                [        -1 / 2, (-np.sqrt(2) + 2) / 4,  (np.sqrt(2) + 2) / 4],
-                [np.sqrt(2) / 2,                -1 / 2,                 1 / 2],
+                [1 / 2, (np.sqrt(2) + 2) / 4, (-np.sqrt(2) + 2) / 4],
+                [-1 / 2, (-np.sqrt(2) + 2) / 4, (np.sqrt(2) + 2) / 4],
+                [np.sqrt(2) / 2, -1 / 2, 1 / 2],
```
Comment (TongguangZhang) [inline tests/unit/test_camera_properties.py:315]: Add formatter ignore

File: tests/unit/test_geolocation.py [hunk 19]
```diff
@@ -637,17 +635,20 @@ class TestGeolocationConvertDetection:
...
         # Setup
         result, expected = detection_in_world.DetectionInWorld.create(
             np.array(
                 [
-                    [  -1.0,   -1.0],
-                    [  -1.0, 3999.0],
-                    [1999.0,   -1.0],
+                    [-1.0, -1.0],
+                    [-1.0, 3999.0],
+                    [1999.0, -1.0],
```
Comment (TongguangZhang) [inline tests/unit/test_geolocation.py:651]: Add formatter ignore

File: tests/unit/test_geolocation.py [hunk 20]
```diff
@@ -681,17 +681,19 @@ def test_normal1(self, detection1: detections_and_time.Detection, affine_matrix:
...
         # Setup
         result, expected = detection_in_world.DetectionInWorld.create(
             np.array(
                 [
-                    [ -1.0,   -1.0],
-                    [ -1.0, 1999.0],
-                    [999.0,   -1.0],
+                    [-1.0, -1.0],
+                    [-1.0, 1999.0],
+                    [999.0, -1.0],
```
Comment (TongguangZhang) [inline tests/unit/test_geolocation.py:696]: Add formatter ignore

File: tests/unit/test_geolocation.py [hunk 21]
```diff
@@ -764,20 +768,20 @@ def test_basic(self,
...
+        result, expected_detection_1 = detection_in_world.DetectionInWorld.create(
             np.array(
                 [
-                    [ 100.0, -100.0],
-                    [ 100.0,  100.0],
+                    [100.0, -100.0],
+                    [100.0, 100.0],
                     [-100.0, -100.0],
-                    [-100.0,  100.0],
+                    [-100.0, 100.0],
```
Comment (TongguangZhang) [inline tests/unit/test_geolocation.py:784]: Add formatter ignore

File: tests/unit/test_geolocation.py [hunk 22]
```diff
@@ -789,15 +793,15 @@ def test_basic(self,
...
             np.array(
                 [
-                    [ 100.0, -100.0],
-                    [ 100.0,    0.0],
-                    [   0.0, -100.0],
-                    [   0.0,    0.0],
+                    [100.0, -100.0],
+                    [100.0, 0.0],
+                    [0.0, -100.0],
+                    [0.0, 0.0],
```
Comment (TongguangZhang) [inline tests/unit/test_geolocation.py:804]: Add formatter ignore

**Review by TongguangZhang — COMMENTED** [review]
Reviewed - Use f-strings and ignore formatter for matrices.

File: documentation/tests/unit/test_add_or_multiply.py (hunk 1 above)
Comment (Xierumeng) [reply documentation/tests/unit/test_add_or_multiply.py:29]: Yes it is.

File: modules/cluster_estimation/cluster_estimation.py (hunk 4 above)
Comment (Xierumeng) [reply modules/cluster_estimation/cluster_estimation.py:324]: Clusters are by index.

File: modules/cluster_estimation/cluster_estimation.py (hunk 3 above)
Comment (Xierumeng) [reply modules/cluster_estimation/cluster_estimation.py:315]: Done.

File: documentation/tests/unit/test_pytest_example.py (hunk 2 above)
Comment (Xierumeng) [reply documentation/tests/unit/test_pytest_example.py:11]: Done.

File: modules/geolocation/camera_properties.py (hunk 9 above)
Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:45]: Done.

File: modules/geolocation/camera_properties.py (hunk 10 above)
Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:54]: Done.

File: modules/geolocation/camera_properties.py (hunk 11 above)
Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:63]: Done.

File: modules/geolocation/camera_properties.py (hunk 12 above)
Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:137]: Done.

File: tests/unit/test_camera_properties.py (hunk 15 above)
Comment (Xierumeng) [reply tests/unit/test_camera_properties.py:225]: Done.

File: tests/unit/test_camera_properties.py (hunk 16 above)
Comment (Xierumeng) [reply tests/unit/test_camera_properties.py:255]: Done.

File: tests/unit/test_camera_properties.py (hunk 17 above)
Comment (Xierumeng) [reply tests/unit/test_camera_properties.py:285]: Done.

File: tests/unit/test_camera_properties.py (hunk 18 above)
Comment (Xierumeng) [reply tests/unit/test_camera_properties.py:315]: Done.

File: tests/unit/test_geolocation.py (hunk 19 above)
Comment (Xierumeng) [reply tests/unit/test_geolocation.py:651]: Done.

File: tests/unit/test_geolocation.py (hunk 20 above)
Comment (Xierumeng) [reply tests/unit/test_geolocation.py:696]: Done.

File: tests/unit/test_geolocation.py (hunk 21 above)
Comment (Xierumeng) [reply tests/unit/test_geolocation.py:784]: Done.

File: tests/unit/test_geolocation.py (hunk 22 above)
Comment (Xierumeng) [reply tests/unit/test_geolocation.py:804]: Done.

File: modules/detections_and_time.py (hunk 6 above)
Comment (Xierumeng) [reply modules/detections_and_time.py:131]: Done.

File: modules/detections_and_time.py (hunk 5 above)
Comment (Xierumeng) [reply modules/detections_and_time.py:71]: Done.

File: modules/drone_odometry_local.py (hunk 7 above)
Comment (Xierumeng) [reply modules/drone_odometry_local.py:47]: Done.

File: modules/drone_odometry_local.py (hunk 8 above)
Comment (Xierumeng) [reply modules/drone_odometry_local.py:104]: Done.

File: modules/drone_odometry_local.py [hunk 23]
```diff
@@ -66,36 +64,43 @@ def create_new(cls,
...
 
         self.orientation = orientation
 
-# pylint: enable=too-few-public-methods
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        # TODO: Update common
+        return f"DroneOrientationLocal (YPR rad): {self.orientation.yaw}, {self.orientation.pitch}, {self.orientation.roll}"
```
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:90]: Line exceeds 100 characters

File: modules/detections_and_time.py [hunk 24]
```diff
@@ -37,61 +35,53 @@ def create(cls,
...
-                + str(self.x1) + " " \
-                + str(self.y1) + " " \
-                + str(self.x2) + " " \
-                + str(self.y2)
-
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        representation = f"cls: {self.label}, conf: {self.confidence}, bounds: {self.x_1} {self.y_1} {self.x_2} {self.y_2}"
```
Comment (TongguangZhang) [inline modules/detections_and_time.py:58]: Keep `__str__()` function consistent with other modules. Either just return the string or always assign `representation = f"...."` and return `representation`

File: modules/detections_and_time.py [hunk 25]
```diff
@@ -114,13 +104,14 @@ def __init__(self, class_private_create_key, timestamp: float):
...
-        representation += "\n" + repr(self.detections)
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        representation = (
+            f"{self.__class__}, time: {int(self.timestamp)}, size: {len(self)}\n"
+            + f"{self.detections}"
+        )
         return representation
```
Comment (TongguangZhang) [inline modules/detections_and_time.py:115]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

File: modules/drone_odometry_local.py [hunk 26]
```diff
@@ -37,24 +33,26 @@ def __init__(self,
         self.east = east
         self.down = down
 
-# pylint: enable=too-few-public-methods
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        return f"DronePositionLocal (NED): {self.north}, {self.east}, {self.down}"
```
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:40]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

File: modules/drone_odometry_local.py (hunk 23 above)
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:90]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

File: modules/drone_odometry_local.py [hunk 27]
```diff
@@ -107,14 +112,22 @@ def create(cls,
...
         assert class_private_create_key is DroneOdometryLocal.__create_key, "Use create() method"
 
         self.position = position
         self.orientation = orientation
+
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        return f"DroneOdometryLocal: {self.position}, {self.orientation}"
```
Comment (TongguangZhang) [inline modules/drone_odometry_local.py:133]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

File: modules/merged_odometry_detections.py [hunk 28]
```diff
@@ -28,17 +28,28 @@ def create(cls,
...
-# pylint: enable=too-few-public-methods
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        representation = (
+            f"Merged: {self.odometry_local}, detections: {len(self.detections)}\n"
+            + f"{self.detections}"
+        )
+        return representation
```
Comment (TongguangZhang) [inline modules/merged_odometry_detections.py:55]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

File: modules/odometry_and_time.py [hunk 29]
```diff
@@ -40,4 +43,11 @@ def __init__(self,
...
 
-# pylint: enable=too-few-public-methods
+    def __str__(self) -> str:
+        """
+        To string.
+        """
+        representation = (
+            f"{self.__class__}, time: {int(self.timestamp)}\n" + f"{self.odometry_data}"
+        )
+        return representation
```
Comment (TongguangZhang) [inline modules/odometry_and_time.py:53]: Keep __str__() function consistent with other modules. Either just return the string or always assign representation = f"...." and return representation

**Review by TongguangZhang — COMMENTED** [review]
Reviewed. Not sure if I'm wrong here but it seems the formatter missed some lines that are over 100 characters?

File: modules/geolocation/camera_properties.py [hunk 30]
```diff
@@ -117,28 +122,30 @@ def create(cls,
...
```
[the remark this answers is not in the document]
Comment (Xierumeng) [reply modules/geolocation/camera_properties.py:143]: This is not a matrix.

**Review by TongguangZhang — APPROVED** [review]
LGTM!

**Pull request merged**